data-raw/units-data.R

```diff
@@ -31,4 +31,22 @@ halounits <- 'Gg'
 unitstable <- rbind(data.frame(variable=hvars, units=hunits),
                     data.frame(variable=haloemis, units=halounits))
 
-devtools::use_data(unitstable, internal=TRUE, overwrite=TRUE)
+###### Get all available hector variables (as defined by data-raw/vars_all.txt)
+vars_all <- read.table("data-raw/vars_all.txt", header=TRUE, as.is=TRUE)
+
+# Split the variables into two lists: Those who can handle a date parameter
+# and those who cannot
+vars_date <- vars_all[vars_all$date_param == TRUE,][,1]
+vars_nodate <- vars_all[vars_all$date_param == FALSE,][,1]
+
+# Convert the strings representing variable names to capability strings.
+vfuns_date <- lapply(vars_date, getFromNamespace, ns='hector')
+vfuns_nodate <- lapply(vars_nodate, getFromNamespace, ns='hector')
+
+# Throws an error if the result of each do.call call is not a character vector
+# of length 1
+vfuns_date <- lapply(vfuns_date, do.call, character(1), args=list())
+vfuns_nodate <- lapply(vfuns_nodate, do.call, character(1), args=list())
```

Comment:
Need to use `vapply` here, not `lapply`.

Reply:
Does `lapply` not check the return value? (I 100% trust your feedback, just curious)

Reply:
`lapply` always returns a list the same length as the thing being mapped over.

```r
result <- lapply(x, FUN, a = a, b = b)
# ...is basically the same as...
result <- list()
for (i in seq(1, length(x)) {
  # In R, lists can hold anything, so this always works
  result[[i]] <- FUN(x[i], a=a, b=b)
}
```

`vapply` takes an _additional_ argument, confusingly called `FUN.VALUE`, that is the same length and type as the expected output of `fun(x[i], ...)`.

```r
# If...
y <- FUN(x[1])
class(y)
#> numeric
length(y)
#> 1
# ...then...
result <- vapply(x, FUN, numeric(1), a = a, b = b)
# ...is roughly the same as...
# Pre-allocate a numeric vector with `length(x)` elements
result <- numeric(length(x))
for (i in seq(1, length(x)){
  r <- FUN(x, a = a, b = b)
  # Check that the output is the specified length and type
  stopifnot(length(r) == 1, class(r) == "numeric")
  result[i] <- r
}
```

Read more by running `?lapply` in the R console (they're both documented on the same page). I also like [this SO answer](https://stackoverflow.com/a/7141669/2477097), though I think it undersells `vapply`. It's true that knowing the output type in advance makes it a bit faster, but I really like the added predictability and explicit error when something unexpected happens.